### PR TITLE
Switch fuzz tests to rules_fuzzing

### PR DESCRIPTION
**Pull request by jonmeow**
This switches to [rules_fuzzing](https://github.com/bazelbuild/rules_fuzzing/) because it has a little more support, and should make it easier to use [oss-fuzz](https://github.com/google/oss-fuzz).

The separate proto-fuzzer feature doesn't seem necessary with the overall changes. However, I think it's using libclang_rt.fuzzer which is built with libstdc++ instead of libc++, so this changes the toolchain to disable libc++ when fuzzer is in use.

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -61,9 +61,10 @@ build:asan --features=asan
 
 # Configuration for enabling LibFuzzer (along with ASan).
 build:fuzzer --features=fuzzer
-
-# Proto fuzzer specific configuration.
-build:proto-fuzzer --features=proto-fuzzer
+build:fuzzer --features=-libc++
+build:fuzzer --@rules_fuzzing//fuzzing:cc_engine=@rules_fuzzing//fuzzing/engines:libfuzzer
```
Comment (chandlerc) [inline .bazelrc:65]: Can we not point this at our build rule for libfuzzer to preserve that?

I'm somewhat sad to lose the really broad consistency of using Clang + libc++ everywhere we can, and having weird differences that crop up because of the oddities of the C++ standard library we happen to be linking...

I guess if this is the only way to make oss-fuzz work, but it seems like ... a really big thing to give up.

File: explorer/fuzzing/BUILD [hunk 2]
```diff
@@ -143,7 +143,10 @@ cc_fuzz_test(
     testonly = 1,
     srcs = ["explorer_fuzzer.cpp"],
     corpus = glob(["fuzzer_corpus/*"]),
-    shard_count = 8,
+    # TODO: This takes several minutes to test the corpus. Either we need to
+    # trim the corpus down to reduce execution time, or find a way to shard
+    # execution.
+    tags = ["manual"],
```
Comment (chandlerc) [inline explorer/fuzzing/BUILD:149]: Worth filing a feature request and linking it here?

File: bazel/cc_toolchains/clang_cc_toolchain_config.bzl [hunk 3]
```diff
@@ -634,6 +642,10 @@ def _impl(ctx):
         ],
     )
 
+    whole_archive_flag = "-Wl,-whole-archive"
+    if ctx.attr.target_cpu in ["darwin", "darwin_arm64"]:
+        whole_archive_flag = "-Wl,-force_load"
```
Comment (chandlerc) [inline bazel/cc_toolchains/clang_cc_toolchain_config.bzl]: Unrelated change, or just something I missed in the description?

Comment (jonmeow) [reply bazel/cc_toolchains/clang_cc_toolchain_config.bzl]: See the macos build error, I'm trying to fix it. I'm guessing this wasn't really used.

File: explorer/fuzzing/BUILD (hunk 2 above)
Comment (jonmeow) [reply explorer/fuzzing/BUILD:149]: Noting that you're objecting to the switch, I'd like to see if we can make progress here before working on a fix. That's why there's a TODO.

File: .bazelrc (hunk 1 above)
Comment (jonmeow) [reply .bazelrc:65]: The outputs are quite a bit different. If you want to maintain separate rules, I would abandon this change, but probably also leave oss-fuzz to someone else to figure out how it could work. The inability to use the provided bazel integration point will increase complexity, and I'm trying to balance how much time I'm spending here. We can always leave oss-fuzz integration for later.

File: bazel/cc_toolchains/clang_cc_toolchain_config.bzl (hunk 3 above)
Comment (chandlerc) [reply bazel/cc_toolchains/clang_cc_toolchain_config.bzl]: Ahh. I see.

I don't think this is quite right sadly... the `-force_load` ld64 flag works differently from the `-whole-archive` one in a big way. It may silently work, but I don't think it'll be doing what is intended.

I think I see how to fix this part of the toolchain config, lemme try it out.

File: .bazelrc (hunk 1 above)
Comment (chandlerc) [reply .bazelrc:65]: I was hoping they were more similar. :sigh: All of the Bazel stuff in `rules_fuzzing` is a bit scary in how complex it has gotten compared to our setup.

That said, I'm not really objecting. I'm mostly nervous. My hope with our build set up was to really minimize the moving pieces we have to support, and this adds a big new one.

But as you point out, this seems likely the only easy path to having OSS-Fuzz working, and this config is primarily useful in that context. So, if this works to get OSS-Fuzz working, it's almost certainly the right call.

Maybe worth leaving a comment that this config isn't supported everywhere and may be brittle, but is designed to roughly match the continuous fuzzing we get from OSS-fuzz?

File: explorer/fuzzing/BUILD (hunk 2 above)
Comment (chandlerc) [reply explorer/fuzzing/BUILD:149]: Sure and this isn't blocking. I'm not suggesting working on a fix ourselves necessarily, upstream may be responsive to a feature request.

File: bazel/cc_toolchains/clang_cc_toolchain_config.bzl (hunk 3 above)
Comment (chandlerc) [reply bazel/cc_toolchains/clang_cc_toolchain_config.bzl]: After reading through some old Java source code that handled this (the first reasonable example I could find) and some of the manual pages, I think something like this would be a fairly principled approach to make this work well with ld64:

https://github.com/chandlerc/carbon-lang/commit/d96a7164150cf62743a0ef19eb12527f322ff5fa

This at least builds for me on macOS (arm64 though).

Happy for you to cherry-pick that in here, or either of us to land it as a separate PR to clean up the flag handling. This

Comment (jonmeow) [reply bazel/cc_toolchains/clang_cc_toolchain_config.bzl]: Trying this out (pushed it separately while I look again at other parts of this)

File: .bazelrc [hunk 4]
```diff
@@ -61,9 +61,10 @@ build:asan --features=asan
...
+build:fuzzer --@rules_fuzzing//fuzzing:cc_engine_instrumentation=libfuzzer
+build:fuzzer --@rules_fuzzing//fuzzing:cc_engine_sanitizer=asan
```
Comment (chandlerc) [inline .bazelrc:67]: Not blocking, just so we don't forget, but if all this works we may want to come back and add UBSan stuff somewhere.

**Review by chandlerc — APPROVED** [review]
Just marking that this LG.

I know based on discussions still some investigation w/ OSS-fuzz, but regardless, if this PR is useful, go for it. Progress seems more important here.

If it ends up not working, may make sense to restore the libfuzzer build bits, but that should be pretty easy.

**Comment by jonmeow** [comment]
I've spent a bit trying to get this to work, so I'll ask for insights.

Here's what I'm trying for an oss-fuzz configuration:
https://github.com/jonmeow/oss-fuzz/pull/new/carbon-lang

Note this branch is used for the carbon-lang repo.

Trying the command:

```
python infra/helper.py build_image $PROJECT_NAME && \
python infra/helper.py build_fuzzers --sanitizer address $PROJECT_NAME && \
python infra/helper.py check_build $PROJECT_NAME
```

([oss-fuzz command reference](https://google.github.io/oss-fuzz/getting-started/new-project-guide/#testing-locally))

I get errors like:

```
ld.lld: error: undefined symbol: lzma_index_end
>>> referenced by elfxx.c:210
>>>               elf64.o:(_Uelf64_get_proc_name_in_image) in archive /lib/x86_64-linux-gnu/libunwind.a
>>> referenced by elfxx.c:210
>>>               elf64.o:(_Uelf64_get_proc_name_in_image) in archive /lib/x86_64-linux-gnu/libunwind.a
```

I've tried passing `-linkopt=-llzma` to bazel and it doesn't seem to change that, although I can't rule out that I have some mistake in build.sh. I've tried disabling static linking, and that just changes the error:

```
/tmp/not-out/tmp8z1c2x8y/numeric_literal_fuzzer: error while loading shared libraries: libunwind.so.8: cannot open shared object file: No such file or directory
```

If I stop trying to install libunwind-dev and just see what happens, it's:

```
ld.lld: error: unable to find library -l:libunwind.so
```

Note, I'm not sure how much of this is due to Carbon's particular toolchain configuration. I do know that rules_cc provides toolchains as well, or bazel has builtin support, but we aren't using either. But TBH I suspect this is more a quirk of how oss-fuzz builds clang, given the particular error. For that, I don't know that I have any great solution available: I can say other systems use bazel with oss-fuzz for C++ code and presumably compile and work, but I'm not gleaning the difference.

**Comment by jonmeow** [comment]
Per discussion, going to abandon this for now. I suspect the oss-fuzz llvm build is related to building llvm from head, but this doesn't seem worth spending more time on right now since we have a separate fuzzing system (unfortunately less visible).
